tiatoolbox/tools/tissuemask.py

```diff
@@ -0,0 +1,225 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software Foundation,
+# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+#
+# This file contains code inspired by StainTools
+# [https://github.com/Peter554/StainTools] written by Peter Byfield.
+#
+# The Original Code is Copyright (C) 2020, TIALab, University of Warwick
+# All rights reserved.
+# ***** END GPL LICENSE BLOCK *****
+
+"""Methods of masking tissue and background."""
+
+from abc import ABC, abstractmethod
+
+import cv2
+import numpy as np
+
+from tiatoolbox.utils.misc import objective_power2mpp
+
+
+class TissueMasker(ABC):
+    """Base class for tissue maskers.
+
+    Takes an image as in put and outputs a mask.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.fitted = False
+
+    @abstractmethod
+    def fit(self, image: np.ndarray, mask=None) -> None:
+        """Fit the masker to the image and given key word parameters.
+
+        Args:
+            image (:class:`numpy.ndarray`): RGB image, usually a WSI thumbnail.
+            mask (:class:`numpy.ndarray`): Target/ground-truth mask.
+        """
+        self.fitted = True
+
+    @abstractmethod
+    def transform(self, image: np.ndarray) -> np.ndarray:
+        """Create and return a tissue mask.
+
+        Args:
+            thumbnail (:class:`numpy.ndarray`): RGB image, usually a WSI thumbnail.
+
+        Returns:
+            np.ndarary: Map of semantic classes spatially over the WSI
+                e.g. regions of tissue vs background.
+        """
+        if not self.fitted:
+            raise Exception("Fit must be called before transform.")
+
+    def fit_transform(self, image: np.ndarray, **fit_params) -> np.ndarray:
+        """Apply :func:`fit` and :func:`transform` in one step.
+
+        Sometimes it can be more optimal to perform both at the same
+        time for a single sample. In this case the base implementation
+        of :fun:`fit` followed by :func:`transform` can be overridden.
+
+        Args:
+            image (:class:`numpy.ndarray`): Image to create mask from.
+            fit_params (dict): Generic key word arguments passed to fit.
+        """
+        self.fit(image, **fit_params)
+        return self.transform(image)
+
+
+class OtsuTissueMasker(TissueMasker):
+    """Tissue masker which uses Otsu's method to determine background."""
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.threshold = None
+
+    def fit(self, image: np.ndarray, mask=None) -> None:
+        # find Otsu's threshold
+        if len(image.shape) == 3:
+            gray = cv2.cvtColor(image, cv2.COLOR_RGB2GRAY)
```

Comment:
Rather than a comment, I think this one should be put into the doc for `OtsuTissueMasker.fit`. 
Personally, I usually get the threshold per image and binarize the image itself using that one instead of a shared one like this. I think the doc will clarify the usage.

Reply:
Also, calling the fit_transform will do what I mention above in case the input is 1 image. Not sure if that is the intended behavior you want.

Reply:
Yes, fit_transform is the behaviour you mentioned of finding the threshold for the image and then binarising using that threshold. I imagine for these basic classes you would only ever use fit_transform. I used this fit and transform structure to try and be consistent with sklearn and staintools. Also if you have some masking method which can be fit to a set of images e.g. a simple random forrest segmenter like this: https://scikit-image.org/docs/dev/user_guide/tutorial_segmentation.html. For example if you had a batch of images and masks, you could fit to that batch then use this to mask other images.